Add config option to filter authn types for credentials

jmyersmsft opened the pull request:
## Bug
Works around NuGet/Home#5651 and NuGet/Home#5286

Currently, it is very difficult to interact with servers that support Negotiate authentication if default credentials are not the right credentials. With dotnet on Linux, it is currently impossible. On Windows, Negotiate itself can work, but if the logged-in Windows user isn't the right user, results can be counter intuitive, and if you actually do need to present a credential specifically via the Basic mechanism (because, for instance, the identity is a special identity defined by the server rather than a real Windows domain user), you're out of luck unless you can write a credential provider (and that option isn't even available with .NET Core NuGet in current released versions)

With .NET Core on Linux, .NET never falls back to Basic if it has a credential for Negotiate, even though it also doesn't try to actually perform Negotiate auth, whether or not an explicit credential is given. I think this is a bug in the .NET framework

## Fix
Details: 
This change extends the mechanism used to filter auth types for credential providers to work with configured credentials as well. The user will specify that `basic` is the only valid authentication type. This will cause NuGet to respond "don't have 
any" when asked for credentials for Negotiate, NTLM, etc, forcing .net to fall back to Basic.

* A property is added to the package source credentials section of the config. `<add key="ValidAuthenticationTypes" value="comma-delimited list of authentication types, or empty to mean 'all authentication types'" />`. The extra property is optional. If the property is not specified, the implied behavior is all, which is backwards compatible.

* A parameter -ValidAuthenticationTypes is added to the nuget.exe sources commands. The parameter can't be independently set. Specifying ValidAuthenticationTypes without UserName and Password is an error. Not specifying ValidAuthenticationTypes when there are auth types in the existing config emits a message and clears the value.

* The syntax for passing credentials in an environment variable has been extended to support the additional ValidAuthenticationTypes parameter. 

* The response contract for the GetCredentials plugin->NuGet command has been tweaked to include an optional AuthenticationTypes property. This should be a backwards-compatible change. If plugins don't understand the new parameter, they're no more broken than they were before.

## Testing/Validation
Tests Added: Yes, a couple of unit tests.
Validation done: Manual test as follows
* NuGet server: On-Premises TFS Package Management, with a feed upstreamed to nuget.org
* Same trivial project used for both environments: `dotnet new console; dotnet add package Newtonsoft.Json`
* Credentials (an arbitrary username and a TFS Personal Access Token) in clear text in nuget.config in project directory
* Tested in Linux Docker container from image microsoft/aspnetcore-build@sha256:1e4043c2484157e029d6b09a3149ebd01c5122f6b62f0a1b6c6f54b90500ea4b.
  * Replaced the NuGet dlls in the dotnet sdk folder
* Tested on Windows with nuget.exe:

jmyersmsft committed:
Add env var to filter auth types for configured creds

Currently, it is very difficult to interact with servers that support
Negotiate authentication if default credentials are not the right
credentials. With dotnet on Linux, it is currently impossible. On
Windows, Negotiate itself can work, but if the logged-in Windows user
isn't the right user, results can be counterintuitive, and if you
actually do need to present credential via the Basic mechanism,
you're out of luck unless you can write a credential provider (and that
option isn't even available with .net core nuget)

I think this is a bug in the library, but with .net core on Linux, .net never falls
back to Basic if it has a credential for Negotiate, even though it also
doesn't try to actually perform Negotiate auth, even if an explicit
credential is given.

This change extends the mechanism used to filter auth types for
credential providers to work with configured credentials as well. The
user will specify NUGET_AUTHENTICATION_TYPES=basic
(NUGET_CREDENTIAL_PROVIDER_OVERRIDE_DEFAULT=true may also be needed for
.net framework-based NuGet to be offered the chance of not using default
credentials). This will cause NuGet to respond "don't have any" when
asked for credentials for Negotiate, NTLM, etc, forcing .net to fall
back to Basic.

jainaashish reviewed (approved):
LGTM

nkolev92 commented:
@jmyersmsft 
As communicated via e-mail. 

We'll be doing a breaking change in https://github.com/NuGet/Home/issues/6961, we're comfortable doing the same thing with this. 
It is where source control comes in the picture :) 


You can either update this PR/open another one or you can wait for us to get on it.

jmyersmsft committed:
Use config instead of env vars

jmyersmsft commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceCredential.cs:
```diff
@@ -42,7 +55,7 @@ public string Password
                     {
                         throw new NuGetConfigurationException(
-                            string.Format(CultureInfo.CurrentCulture, Resources.UnsupportedDecryptPassword, Source), e);
+                            String.Format(CultureInfo.CurrentCulture, Resources.UnsupportedDecryptPassword, Source), e);
```
Gah, I think ReSharper did that. I'll fix it in the morning.

jmyersmsft replied:
I'll add some more comments in this file in the morning too.

jmyersmsft commented:
@nkolev92 I've updated my branch.

This push is a fairly substantial change from the original, switching to obtain the filter list from the config instead of the environment.

* I added a property to the credential section of the config: `<add key="ValidAuthenticationTypes" value="comma-delimited list of authentication types, or empty or * to mean 'all authentication types'" />`
  * `*` is allowed so that the user has a way to explicitly set to all types in the nuget sources command (if the user doesn't specify the parameter, any existing value is retained)
* I added a parameter `-ValidAuthenticationTypes` to `nuget sources`. It can be set independently of `-Username` and `-Password`, but the user must either pass those or (for update) already have them configured.
* I extended the syntax for passing credentials in environment variables to additionally support a `ValidAuthenticationTypes` parameter
* I also tweaked the response contract for the `GetCredentials` plugin->NuGet command to include an optional `AuthenticationTypes` property. This should be a backwards-compatible change. If plugins don't understand the new parameter, they're no more broken than they were before.

jmyersmsft committed:
Add comments and fix resharper goof

nkolev92 commented:
@PatoBeltran 
There are some config changes here about which you'd have good context.

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx:
```diff
@@ -5476,4 +5476,8 @@ This option should be used when specifying the certificate via -CertificateSubje
     <value>Does not append "api/v2/packages" to the source URL.</value>
   </data>
+  <data name="SourcesCommandValidAuthenticationTypesDescription" xml:space="preserve">
+    <value>Comma-separated list of valid authentication types for this source, or * to allow all authentication types, which is the default. Example: basic,negotiate</value>
```
Don't think we should support `*`, nothing should imply all.

jmyersmsft replied:
How do you specify on the command line that you want to *change* the authentication types from a specific value to all, while also allowing the user to leave the authentication types untouched while updating the username and password?

nkolev92 replied:
yeah, I get your point....//cc @anangaur, @rrelyea let's discuss this.

nkolev92 replied:
Discussed this with @rrelyea and @rohit21agrawal offline. 

I don't think we care about that scenario. 
We don't think that the authentication types should be preserved if someone does update. 

Let's not preserve the authentication types. 

Basically running
`nuget sources update -name nuget.org -user user -pass pass -validauthenticationtypes basic`
`nuget sources update -name nuget.org -user user2 -pass pass2 `

should result with user2 and pass2 in the config only. 

We can add a warning indicating when a user it updating the credentials and we are about to drop the valid authentication types value. 

Basically, let's not do `*`

nkolev92 commented on src/NuGet.Clients/NuGet.Credentials/PluginCredentialProvider.cs:
```diff
@@ -124,11 +125,9 @@ public PluginCredentialProvider(Common.ILogger logger, string path, int timeoutS
-                    {
-                        result = new AuthTypeFilteredCredentials(result, response.AuthTypes);
-                    }
+                    var result = new AuthTypeFilteredCredentials(
```
Why change this?

It has no functional impact as far as I can tell.

jmyersmsft replied:
It's not PluginCredentialProvider's responsibility to figure out what the filter means

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/SettingsCredentialProvider.cs:
```diff
@@ -30,21 +31,20 @@ public SettingsCredentialProvider(Configuration.IPackageSourceProvider packageSo
-            NetworkCredential credentials;
             // If we are retrying, the stored credentials must be invalid.
-            if (!retrying && (credentialType == CoreV2.NuGet.CredentialType.RequestCredentials) && TryGetCredentials(uri, out credentials))
+            if (!retrying && (credentialType == CoreV2.NuGet.CredentialType.RequestCredentials) && TryGetCredentials(uri, out var credentials, out var username))
```
Do we really need to pass back the username? 
Why not just use it from the Credentials object?

jmyersmsft replied:
ICredentials doesn't directly expose the username

nkolev92 replied:
ah, missed the fact that you're changing it to ICredentials.

nkolev92 commented on src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs:
```diff
@@ -199,11 +200,9 @@ private static CredentialResponse GetAuthenticationCredentialsResponseToCredenti
-                {
-                    result = new AuthTypeFilteredCredentials(result, credentialResponse.AuthenticationTypes);
-                }
+                ICredentials result = new AuthTypeFilteredCredentials(
```
Likewise...I think this is an unnecesarry change as it has no functional impact.

nkolev92 commented on src/NuGet.Core/NuGet.Common/AuthTypeFilteredCredentials.cs:
```diff
@@ -3,45 +3,44 @@
     {
-        private readonly IList<string> _authTypes;
-        private readonly ICredentials _innerCredential;
+        public IReadOnlyList<string> AuthTypes { get; }
```
Why remove readonly?

jmyersmsft replied:
Turned it into a read-only property instead of a readonly field. The backing field of a read-only auto-property is readonly.

nkolev92 commented on src/NuGet.Core/NuGet.Common/AuthTypeFilteredCredentials.cs:
```diff
@@ -3,45 +3,44 @@
         {
-            return _authTypes.Any(x => StringComparer.OrdinalIgnoreCase.Equals(x, authType))
-                ? _innerCredential.GetCredential(uri, authType)
+            return authType == null || !AuthTypes.Any() || AuthTypes.Any(x => StringComparer.OrdinalIgnoreCase.Equals(x, authType))
```
Why change this?

jmyersmsft replied:
NuGet has elsewhere assumed that you can call GetCredential with an authType of null, and the null comparison allows that to pass through. The !AuthTypes.Any() check moves the responsibility of determining that "an empty filter means allow all" from the caller where it doesn't belong, to this class, where it does.

nkolev92 replied:
ok, makes sense.

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetCredentialsResponse.cs:
```diff
@@ -28,6 +29,9 @@ public sealed class GetCredentialsResponse
         /// </summary>
         public string Username { get; }
 
+        public IReadOnlyList<string> AuthenticationTypes { get; }
```
I get that this would not break the serialization but it feels very weird. 
We'd need to call it out explicitly in the spec. 

//cc @dtivel 
Can you please take a look at this?

jmyersmsft replied:
How is this weird? This is necessary to propagate the authentication information NuGet has, and it mirrors the communication going the other direction.

nkolev92 replied:
It's not the fact that it's necesarry that's weird. 

It's the fact that plugins built against the V1 (4.3.0) protocol would work without a hitch even though a different thing is being sent over the network, just because MissingMemberHandling is not error during the deserialization.

Ideally every change is accompanied by a protocol change.

nkolev92 reviewed (commented):
A couple of small comments and suggestions. 
//cc @NuGet/nuget-client can someone else take a look?

> * I added a property to the credential section of the config: `<add key="ValidAuthenticationTypes" value="comma-delimited list of authentication types, or empty or * to mean 'all authentication types'" />`
>  * `*` is allowed so that the user has a way to explicitly set to all types in the nuget sources command (if the user doesn't specify the parameter, any existing value is retained)

Why `*`? I don't think there's a need for that, `all` is the implied value. 

> * I added a parameter `-ValidAuthenticationTypes` to `nuget sources`. It can be set independently of `-Username` and `-Password`, but the user must either pass those or (for update) already have them configured.

I think this is reasonable from an experience standpoint... @anangaur can you please look into this. 

> * I extended the syntax for passing credentials in environment variables to additionally support a `ValidAuthenticationTypes` parameter
That's cool.

> * I also tweaked the response contract for the `GetCredentials` plugin->NuGet command to include an optional `AuthenticationTypes` property. This should be a backwards-compatible change. If plugins don't understand the new parameter, they're no more broken than they were before.

I get this and it's fine but not a huge fan :)

nkolev92 commented:
@jmyersmsft
Overall looks good, minor thing, please remove the `*` and use empty as an implication of all. 
Don't preserve the ValidAuthenticationTypes when doing update. 

I'll reverse engineer a spec describing the behavior changes that will help with the docs/release notes.

jmyersmsft committed:
PR feedback; remove *

jmyersmsft commented:
`*` is removed. Specifying ValidAuthenticationTypes without UserName and Password is now an error. Not specifying ValidAuthenticationTypes when there are auth types in the existing config emits a message and clears the value.

nkolev92 reviewed (approved):
LGTM. 

I have amended the original PR description to closer describe what the experience change is. 

@NuGet/nuget-client Can I get another reviewer?
@dtivel I'd love to get your opinion on the GetCredentialsResponse changes.

nkolev92 commented:
@jmyersmsft Can you please rebase from dev?

jmyersmsft commented:
@nkolev92 rebased as requested.

jainaashish commented on src/NuGet.Core/NuGet.Protocol/Plugins/RequestHandlers/GetCredentialsRequestHandler.cs:
```diff
@@ -154,19 +155,29 @@ public void AddOrUpdateSourceRepository(SourceRepository sourceRepository)
+                }
+                else
+                {
+                    networkCredential = credential?.GetCredential(packageSource.SourceUri, null);
```
Just curious what are the cases where `credential` is neither `NetworkCredential` nor `AuthTypeFilteredCredential`?

nkolev92 replied:
edit: nvm...this is fine :) 
```
using System;

public class Program
{
    public static void Main()
    {
        object[] items = { "bla", null };
        foreach (var item in items)
        {
            if (item is string obj)
                Console.WriteLine($"Type: {obj.GetType().Name}, Value: {obj}");
        }
    }
}
```

@jainaashish This would be for the null case only. considering the call returns a Task<NetworkCredential>

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/Plugins/RequestHandlers/GetCredentialsRequestHandler.cs:
```diff
@@ -154,19 +155,29 @@ public void AddOrUpdateSourceRepository(SourceRepository sourceRepository)
                 }
 
-                if (credential == null)
+                if (credential is AuthTypeFilteredCredentials filteredCredentials)
```
Actually this would throw in the case of null. 

An equivalent example would be the following: 
```
using System;

public class Program
{
    public static void Main()
    {
        object[] items = { "bla", null };
        foreach (var item in items)
        {
            if (item is var obj)
                Console.WriteLine($"Type: {obj.GetType().Name}, Value: {obj}");
        }
    }
}
```
